### PR TITLE
[aggregator] Add role tag to aggregator error metrics

### PR DESCRIPTION
**Pull request by vpranckaitis**
**What this PR does / why we need it**:

Adds role tag to aggregator's `AddUntimed`, `AddTimed`, `AddPassthrough` and `AddForwarded` error metrics

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/aggregator/aggregator/aggregator.go [hunk 1]
```diff
@@ -799,45 +809,78 @@ func (m *aggregatorAddMetricMetrics) ReportError(err error) {
...
+	)
 	return aggregatorAddUntimedMetrics{
-		aggregatorAddMetricMetrics: newAggregatorAddMetricMetrics(scope, opts),
-		invalidMetricTypes: scope.Tagged(map[string]string{
-			"reason": "invalid-metric-types",
-		}).Counter("errors"),
+		aggregatorAddMetricSuccessMetrics: newAggregatorAddMetricSuccessMetrics(scope, opts),
+
+		leaderErrors:    newAggregatorAddUntimedErrorMetrics(leaderScope),
+		nonLeaderErrors: newAggregatorAddUntimedErrorMetrics(nonLeaderScope),
```
Comment (linasm) [inline src/aggregator/aggregator/aggregator.go:852]: Nit: I suggest reducing the boilerplate here (and in other places) by moving to something like this:
```suggestion
		leaderErrors:    newAggregatorAddUntimedErrorMetrics(withRole("leader", scope)),
		nonLeaderErrors: newAggregatorAddUntimedErrorMetrics(withRole("non-leader", scope)),
```

File: src/aggregator/aggregator/aggregator.go [hunk 2]
```diff
@@ -799,45 +809,78 @@ func (m *aggregatorAddMetricMetrics) ReportError(err error) {
...
 
-func (m *aggregatorAddUntimedMetrics) ReportError(err error) {
+func (m *aggregatorAddUntimedMetrics) ReportError(err error, role ElectionState) {
+	var errors *aggregatorAddUntimedErrorMetrics
+	switch role {
+	case LeaderState:
+		errors = &m.leaderErrors
+	default:
+		errors = &m.nonLeaderErrors
+	}
```
Comment (linasm) [inline src/aggregator/aggregator/aggregator.go:863]: Nit - consider simplifying to (here and in other places):
```suggestion
	errors := &m.nonLeaderErrors
	if role == LeaderState {
		errors = &m.leaderErrors
	}
```

File: src/aggregator/aggregator/aggregator_test.go [hunk 3]
```diff
@@ -984,34 +984,44 @@ func TestAggregatorOwnedShards(t *testing.T) {
 	}
 }
 
-func TestAggregatorAddMetricMetrics(t *testing.T) {
+func TestAggregatorAddUntimedMetrics(t *testing.T) {
```
Comment (linasm) [inline src/aggregator/aggregator/aggregator_test.go:987]: I know this is more a comment towards the original code, but would it be worth adding similar test cases for passthrough and forwarded metrics as well?

File: src/aggregator/aggregator/aggregator_test.go [hunk 4]
```diff
@@ -984,34 +984,44 @@ func TestAggregatorOwnedShards(t *testing.T) {
...
-	m.ReportError(errWriteValueRateLimitExceeded)
-	m.ReportError(xerrors.NewRenamedError(errArrivedTooLate, fmt.Errorf("errorrr")))
-	m.ReportError(errors.New("foo"))
+	for _, state := range []ElectionState{LeaderState, FollowerState} {
+		m.ReportError(errInvalidMetricType, state)
+		m.ReportError(errShardNotOwned, state)
+		m.ReportError(errAggregatorShardNotWriteable, state)
+		m.ReportError(errWriteNewMetricRateLimitExceeded, state)
+		m.ReportError(errWriteValueRateLimitExceeded, state)
+		m.ReportError(xerrors.NewRenamedError(errArrivedTooLate, fmt.Errorf("errorrr")), state)
```
Comment (linasm) [inline src/aggregator/aggregator/aggregator_test.go:998]: Nit - you can get rid of that `fmt` import by using `errors.New` instead:
```suggestion
		m.ReportError(xerrors.NewRenamedError(errArrivedTooLate, errors.New("errorrr")), state)
```

File: src/aggregator/aggregator/aggregator_test.go [hunk 5]
```diff
@@ -1036,31 +1046,42 @@ func TestAggregatorAddTimedMetrics(t *testing.T) {
...
-	m.ReportError(xerrors.NewRenamedError(errArrivedTooLate, fmt.Errorf("errorrr")))
-	m.ReportError(errors.New("foo"))
+	for _, state := range []ElectionState{LeaderState, FollowerState} {
+		m.ReportError(errShardNotOwned, state)
+		m.ReportError(errAggregatorShardNotWriteable, state)
+		m.ReportError(errWriteNewMetricRateLimitExceeded, state)
+		m.ReportError(errWriteValueRateLimitExceeded, state)
+		m.ReportError(errTooFarInTheFuture, state)
+		m.ReportError(errTooFarInThePast, state)
+		m.ReportError(xerrors.NewRenamedError(errArrivedTooLate, fmt.Errorf("errorrr")), state)
```
Comment (linasm) [inline src/aggregator/aggregator/aggregator_test.go:1056]: Same nit.

File: src/aggregator/aggregator/aggregator.go (hunk 1 above)
Comment (vpranckaitis) [reply src/aggregator/aggregator/aggregator.go]: https://github.com/m3db/m3/pull/3559/commits/6d01251ee123ea551f405228b99f8937b074f838

File: src/aggregator/aggregator/aggregator.go (hunk 2 above)
Comment (vpranckaitis) [reply src/aggregator/aggregator/aggregator.go]: https://github.com/m3db/m3/pull/3559/commits/f7dac96dc1a995ee2d63a12ab7e657a51928138e

File: src/aggregator/aggregator/aggregator_test.go (hunk 3 above)
Comment (vpranckaitis) [reply src/aggregator/aggregator/aggregator_test.go:986]: https://github.com/m3db/m3/pull/3559/commits/f7b68c2758f72dcf2b13d1eb4ed6b71d2c11fd63

File: src/aggregator/aggregator/aggregator_test.go (hunk 4 above)
Comment (vpranckaitis) [reply src/aggregator/aggregator/aggregator_test.go]: https://github.com/m3db/m3/pull/3559/commits/63d14798b89bf8d027377e9ff0c30069df16c2c4

File: src/aggregator/aggregator/aggregator_test.go (hunk 5 above)
Comment (vpranckaitis) [reply src/aggregator/aggregator/aggregator_test.go]: https://github.com/m3db/m3/pull/3559/commits/63d14798b89bf8d027377e9ff0c30069df16c2c4

**Review by linasm — APPROVED** [review]
LGTM

**Pull request merged**